README
B::Debug - Walk Perl syntax tree, printing debug info about ops
	   A simple list version of B::Terse, i.e. B::Concise

SYNOPSIS
            perl -MO=Debug foo.pl
            perl -MO=Debug,-exec foo.pl

With option -exec, walks tree in execute order, otherwise in basic
order.

Copyright (c) 1996, 1997 Malcolm Beattie
Copyright (c) 2008 Reini Urban

	This program is free software; you can redistribute it and/or modify
	it under the terms of either:

	a) the GNU General Public License as published by the Free
	Software Foundation; either version 1, or (at your option) any
	later version, or

	b) the "Artistic License" which comes with this kit.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See either
    the GNU General Public License or the Artistic License for more details.

    You should have received a copy of the Artistic License with this kit,
    in the file named "Artistic".  If not, you can get one from the Perl
    distribution. You should also have received a copy of the GNU General
    Public License, in the file named "Copying". If not, you can get one
    from the Perl distribution or else write to the Free Software Foundation,
    Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA.



Changes
-*- change-log -*-

1.23 2015-01-25 rurban
        * again being hit by EUMM not being able to do parallel cpansign

1.22 2014-10-12 rurban
	* add 5.21.5 support: changed test, from blead for split optim.,
	  added METHOP
	* print 0x00000000 as 0x0
	* add -d -MOd=Debug support
	* README: fixed copyright date

1.21 2014-08-05 rurban
	* add missing META.json to MANIFEST (RT#97722 by KENTNL)

1.20 2014-08-03 rurban
	* avoid exists on an array element (aaron crane)

1.19 2013-11-07 rurban
	* install into site (again) since 5.12
        * support lexical subs, NAME instead of GV.
        * fix double SV FLAGS
        * fix wrong xpv_cur, add xpv_len

1.18 2012-08-17 rurban
	* support new B::PADLIST (experimentally on blead)

1.17 2011-11-25 rurban
	* FSF address change

1.16 2010-09-16 rurban
	* fix test 9 on Win32, shell quoting

1.15 2010-09-12 rurban
	* fix sv->GvFLAGS => gv->GvFLAGS
	* fixes for 5.8.0 (COP->io) and 5.6.1 (GV->SAFENAME)
	* fix $have_B_Flags_extra warning

1.14 2010-09-11 rurban
	* Remove qw(...) as parentheses deprecated warnings.
	  (PERL#77758 by Paul Johnson)

1.13 2010-09-09 rurban
	* print name of op_type
	* print ppaddr consistent with other op addr
	* fix cop_io
	* omit cv->OUTSIDE_SEQ for 5.6
	* fix NULL specials
	* fix NV assertion for CV
	* stabilize tests for space in runperl path
	* fix t/debug.t test 7

1.12 2010-02-10 rurban
	* remove archlib installation cruft, and use the proper PM rule.
	By Todd Rinaldo (toddr)

1.11 2008-07-14 rurban
	* avoid B::Flags in CORE tests not to crash on old XS in @INC

1.10 2008-06-28 rurban
	* require 5.006; Test::More not possible in 5.00505
	* our => my

1.09 2008-06-18 rurban
	* minor META.yml syntax fix
	* 5.8.0 ending nextstate test failure: be more tolerant
	* PREREQ_PM Test::More

1.08 2008-06-17 rurban
	* support 5.00558 - 5.6.2

1.07 2008-06-16 rurban
	* debug.t: fix strawberry perl quoting issue

1.06 2008-06-15 rurban
	* added B::Flags output
	* dual-life CPAN as B-Debug-1.06 and CORE. Removed from B-C-1.04_20
	* protect scalar(@array) if tied arrays leave out FETCHSIZE
	* special Makefile INST_ARCHLIB hack to install into ARCHLIB for
	the default perl @INC searchorder

1.05_03 2008-04-16 rurban
	* ithread fixes in B::AV
	* B-C-1.04_15

B-C-1.04_09 2008-02-24 rurban
	* support 5.8 (import Asmdata)

1.05_02 2008-02-21 rurban
	* added _printop
	* B-C-1.04_08 and CORE

1.05_01 2008-02-05 rurban
	* 5.10 fix for op->seq
	* B-C-1.04_04 

===================================================================
Maybe Relevant Changes from CORE

For the exact Changes see
  http://public.activestate.com/cgi-bin/perlbrowse/b/ext/B/B/Debug.pm
  http://public.activestate.com/cgi-bin/perlbrowse/b/ext/B/t/debug.t

Gurusamy Sarathy <gsar@cpan.org>  2000-12-17 22:49:13
    Subject: [ID 20000928.002] perlcc & ByteCode.pm option mismatch
    Did not apply cleanly, manual intervention was needed.

    Subject: [ID 20001003.006] B::Debug not -w clean

    Test harness update to sync with the new perlcc,
    from Simon Cozens.

Nick Ing-Simmons <nick@ing-simmons.net>  1999-01-02 10:04:02
    Integrate ext/B changes from //depot/cfgperl
    'copy in' ext/B/B.xs
    	(@2460..) 'merge in' ext/B/B/C.pm ext/B/B/CC.pm
    	ext/B/B/Stackobj.pm (@2524..)

Nick Ing-Simmons <nick@ing-simmons.net>  1999-01-02 10:04:02
    @2491 Integrate ext/B changes from //depot/cfgperl

Nick Ing-Simmons <nick@ing-simmons.net>  1998-12-20 14:21:29
    Save _all_ GV's which have SV, AV or HV set.

Gurusamy Sarathy <gsar@cpan.org>  1998-07-22 02:08:00
    fix up B modules for PL_* changes

Malcolm Beattie <mbeattie@sable.ox.ac.uk>  1998-07-21 18:13:16
    Compiler docs for 5.005
    Message-Id: <199807211713.SAA20735@sable.ox.ac.uk>

Gurusamy Sarathy <gsar@cpan.org>  1998-07-22 01:29:09
    s/PL_sv/PL_bytecode_sv/ etc., so we have unique, case-insensitive
    names

Gurusamy Sarathy <gsar@cpan.org>  1998-07-21 07:12:00
    fix bytecode.pl with moved var names

Gurusamy Sarathy <gsar@cpan.org>  1998-07-20 09:38:39
    complete s/foo/PL_foo/ changes (all escaped cases identified with
    brute force search script).  Result builds and passes all tests on
    Solaris.  win32 and PERL_OBJECT are still untested.

Nick Ing-Simmons <nick@ing-simmons.net>  1998-07-18 13:53:03
    PL_ prefix to all perlvars, part1
       Builds and passes all tests at one limit i.e. -DPERL_GLOBAL_STRUCT

Ilya Zakharevich <ilya@math.ohio-state.edu>  1998-07-15 06:10:36
    Minor improvements to perlcc
    Message-Id: <199807151010.GAA11270@monk.mps.ohio-state.edu>

Gurusamy Sarathy <gsar@cpan.org>  1998-07-15 10:01:41
    add stub docs for ext/B, other minor tweaks

Gurusamy Sarathy <gsar@cpan.org>  1998-07-09 05:37:48
    get it building again on win32

Stephen McCamant <alias@mcs.com>  1998-06-22 21:19:43
    Inheritance of B:: classes
    Message-Id: <m0yoIgR-000EP2C@alias-2.pr.mcs.net>

Ed Peschko <epeschko@den-mdev1>  1998-04-29 21:02:36
    [ PATCH 5.004_64 ] Integrated regression tests for compiler
    added perlcc.PL

Charles Bailey <bailey@newman.upenn.edu>  1998-03-02 01:48:27
	Miscellaneous minor fixes

Spider Boardman <spider@orb.nashua.nh.us>  1998-03-01 02:02:47
	Almost OK: 5.004_61 (threads, perlio)

Gurusamy Sarathy <gsar@cpan.org>  1998-02-26 03:56:19
    [win32] various cleanups so that B can be built as "just another extension"
     - export symbols needed for building B
     - bset_obj_store() is needed by byterun(), so define it there instead
       of at B.xs, and export it
     - freadpv() is only used in B.xs, so move it there
     - byte*.h are now included by perl.h
     - regenerate embed*.h

Gurusamy Sarathy <gsar@cpan.org>  1998-02-22 02:40:56
    [win32] get compiler building under win32 (needed Makefile.PL
    hacks that could be applicable to other platforms)

Malcolm Beattie <mbeattie@sable.ox.ac.uk>  1998-02-25 17:44:34
	More compiler tweaks.

Malcolm Beattie <mbeattie@sable.ox.ac.uk>  1998-02-20 18:23:47

Makefile.PL
use ExtUtils::MakeMaker;
require 5.006;

WriteMakefile
  (
   NAME	           => "B::Debug",
   VERSION_FROM    => "Debug.pm",
   PREREQ_PM => {'B'         => 0,
                 'Test::More' => 0},
   ($] >= 5.005 ?
    ('AUTHOR' => 'Malcolm Beattie <mbeattie@sable.ox.ac.uk>, Reini Urban <rurban@cpan.org>',
     'ABSTRACT' => 'print debug info about ops') : ()),
   ($ExtUtils::MakeMaker::VERSION gt '6.30' ?
    ('LICENSE'                => 'perl', ) : ()),
   (($ExtUtils::MakeMaker::VERSION gt '6.31' and $ExtUtils::MakeMaker::VERSION lt '6.46') ?
    ('EXTRA_META'  => "recommends:\n" .
     "    B::Flags:                      0.04\n"
    ) : ()),
   ($ExtUtils::MakeMaker::VERSION gt '6.46' ?
    ('META_MERGE'  => {"recommends" => {
                       "B::Flags" => "0.04"}}
    ) : ()),
   # Part of core perl since 5.005005, but core wants it to be in site since 5.12
   'INSTALLDIRS' => ( $] > 5.005005 && $] < 5.011 ? "perl" : "site" ),
   # install into ARCHLIB
   ($] < 5.011) ? (PM => {
       'Debug.pm' => '$(INST_ARCHLIB)/B/Debug.pm',
   }) : (),
   SIGN => 1,
  );


Debug.pm
package B::Debug;

our $VERSION = '1.19';

use strict;
require 5.006;
use B qw(peekop class walkoptree walkoptree_exec
         main_start main_root cstring sv_undef SVf_NOK SVf_IOK);
use Config;
my (@optype, @specialsv_name);
require B;
if ($] < 5.009) {
  require B::Asmdata;
  B::Asmdata->import (qw(@optype @specialsv_name));
} else {
  B->import (qw(@optype @specialsv_name));
}

if ($] < 5.006002) {
  eval q|sub B::GV::SAFENAME {
    my $name = (shift())->NAME;
    # The regex below corresponds to the isCONTROLVAR macro from toke.c
    $name =~ s/^([\cA-\cZ\c\\c[\c]\c?\c_\c^])/"^".chr(64 ^ ord($1))/e;
    return $name;
  }|;
}

my ($have_B_Flags, $have_B_Flags_extra);
if (!$ENV{PERL_CORE}){ # avoid CORE test crashes
  eval { require B::Flags and $have_B_Flags++ };
  $have_B_Flags_extra++ if $have_B_Flags and $B::Flags::VERSION gt '0.03';
}
my %done_gv;

sub _printop {
  my $op = shift;
  my $addr = ${$op} ? $op->ppaddr : '';
  $addr =~ s/^PL_ppaddr// if $addr;
  return sprintf "0x%08x %6s %s", ${$op}, ${$op} ? class($op) : '', $addr;
}

sub B::OP::debug {
    my ($op) = @_;
    printf <<'EOT', class($op), $$op, _printop($op), _printop($op->next), _printop($op->sibling), $op->targ, $op->type, $op->name;
%s (0x%lx)
	op_ppaddr	%s
	op_next		%s
	op_sibling	%s
	op_targ		%d
	op_type		%d	%s
EOT
    if ($] > 5.009) {
	printf <<'EOT', $op->opt;
	op_opt		%d
EOT
    } else {
	printf <<'EOT', $op->seq;
	op_seq		%d
EOT
    }
    if ($have_B_Flags) {
        printf <<'EOT', $op->flags, $op->flagspv, $op->private, $op->privatepv;
	op_flags	%d	%s
	op_private	%d	%s
EOT
    } else {
        printf <<'EOT', $op->flags, $op->private;
	op_flags	%d
	op_private	%d
EOT
    }
}

sub B::UNOP::debug {
    my ($op) = @_;
    $op->B::OP::debug();
    printf "\top_first\t%s\n", _printop($op->first);
}

sub B::BINOP::debug {
    my ($op) = @_;
    $op->B::UNOP::debug();
    printf "\top_last \t%s\n", _printop($op->last);
}

sub B::LOOP::debug {
    my ($op) = @_;
    $op->B::BINOP::debug();
    printf <<'EOT', _printop($op->redoop), _printop($op->nextop), _printop($op->lastop);
	op_redoop	%s
	op_nextop	%s
	op_lastop	%s
EOT
}

sub B::LOGOP::debug {
    my ($op) = @_;
    $op->B::UNOP::debug();
    printf "\top_other\t%s\n", _printop($op->other);
}

sub B::LISTOP::debug {
    my ($op) = @_;
    $op->B::BINOP::debug();
    printf "\top_children\t%d\n", $op->children;
}

sub B::PMOP::debug {
    my ($op) = @_;
    $op->B::LISTOP::debug();
    printf "\top_pmreplroot\t0x%x\n", $] < 5.008 ? ${$op->pmreplroot} : $op->pmreplroot;
    printf "\top_pmreplstart\t0x%x\n", ${$op->pmreplstart};
    printf "\top_pmnext\t0x%x\n", ${$op->pmnext} if $] < 5.009005;
    if ($Config{'useithreads'}) {
      printf "\top_pmstashpv\t%s\n", cstring($op->pmstashpv);
      printf "\top_pmoffset\t%d\n", $op->pmoffset;
    } else {
      printf "\top_pmstash\t%s\n", cstring($op->pmstash);
    }
    printf "\top_precomp\t%s\n", cstring($op->precomp);
    printf "\top_pmflags\t0x%x\n", $op->pmflags;
    printf "\top_reflags\t0x%x\n", $op->reflags if $] >= 5.009;
    printf "\top_pmpermflags\t0x%x\n", $op->pmpermflags if $] < 5.009;
    printf "\top_pmdynflags\t0x%x\n", $op->pmdynflags if $] < 5.009;
    $op->pmreplroot->debug if $] < 5.008;
}

sub B::COP::debug {
    my ($op) = @_;
    $op->B::OP::debug();
    my $warnings = ref $op->warnings ? ${$op->warnings} : 0;
    printf <<'EOT', $op->label, $op->stashpv, $op->file, $op->cop_seq, $op->arybase, $op->line, $warnings;
	cop_label	"%s"
	cop_stashpv	"%s"
	cop_file	"%s"
	cop_seq		%d
	cop_arybase	%d
	cop_line	%d
	cop_warnings	0x%x
EOT
  if ($] > 5.008 and $] < 5.011) {
    my $cop_io = class($op->io) eq 'SPECIAL' ? '' : $op->io->as_string;
    printf("	cop_io		%s\n", cstring($cop_io));
  }
}

sub B::SVOP::debug {
    my ($op) = @_;
    $op->B::OP::debug();
    printf "\top_sv\t\t0x%x\n", ${$op->sv};
    $op->sv->debug;
}

sub B::PVOP::debug {
    my ($op) = @_;
    $op->B::OP::debug();
    printf "\top_pv\t\t%s\n", cstring($op->pv);
}

sub B::PADOP::debug {
    my ($op) = @_;
    $op->B::OP::debug();
    printf "\top_padix\t%ld\n", $op->padix;
}

sub B::NULL::debug {
    my ($sv) = @_;
    if ($$sv == ${sv_undef()}) {
	print "&sv_undef\n";
    } else {
	printf "NULL (0x%x)\n", $$sv;
    }
}

sub B::SV::debug {
    my ($sv) = @_;
    if (!$$sv) {
	print class($sv), " = NULL\n";
	return;
    }
    printf <<'EOT', class($sv), $$sv, $sv->REFCNT;
%s (0x%x)
	REFCNT		%d
EOT
    printf "\tFLAGS\t\t0x%x", $sv->FLAGS;
    if ($have_B_Flags) {
      printf "\t%s", $have_B_Flags_extra ? $sv->flagspv(0) : $sv->flagspv;
    }
    print "\n";
}

sub B::RV::debug {
    my ($rv) = @_;
    B::SV::debug($rv);
    printf <<'EOT', ${$rv->RV};
	RV		0x%x
EOT
    $rv->RV->debug;
}

sub B::PV::debug {
    my ($sv) = @_;
    $sv->B::SV::debug();
    my $pv = $sv->PV();
    printf <<'EOT', cstring($pv), $sv->CUR, $sv->LEN;
	xpv_pv		%s
	xpv_cur		%d
	xpv_len		%d
EOT
}

sub B::IV::debug {
    my ($sv) = @_;
    $sv->B::SV::debug();
    printf "\txiv_iv\t\t%d\n", $sv->IV if $sv->FLAGS & SVf_IOK;
}

sub B::NV::debug {
    my ($sv) = @_;
    $sv->B::IV::debug();
    printf "\txnv_nv\t\t%s\n", $sv->NV if $sv->FLAGS & SVf_NOK;
}

sub B::PVIV::debug {
    my ($sv) = @_;
    $sv->B::PV::debug();
    printf "\txiv_iv\t\t%d\n", $sv->IV if $sv->FLAGS & SVf_IOK;
}

sub B::PVNV::debug {
    my ($sv) = @_;
    $sv->B::PVIV::debug();
    printf "\txnv_nv\t\t%s\n", $sv->NV if $sv->FLAGS & SVf_NOK;
}

sub B::PVLV::debug {
    my ($sv) = @_;
    $sv->B::PVNV::debug();
    printf "\txlv_targoff\t%d\n", $sv->TARGOFF;
    printf "\txlv_targlen\t%u\n", $sv->TARGLEN;
    printf "\txlv_type\t%s\n", cstring(chr($sv->TYPE));
}

sub B::BM::debug {
    my ($sv) = @_;
    $sv->B::PVNV::debug();
    printf "\txbm_useful\t%d\n", $sv->USEFUL;
    printf "\txbm_previous\t%u\n", $sv->PREVIOUS;
    printf "\txbm_rare\t%s\n", cstring(chr($sv->RARE));
}

sub B::CV::debug {
    my ($sv) = @_;
    $sv->B::PVNV::debug();
    my ($stash) = $sv->STASH;
    my ($start) = $sv->START;
    my ($root)  = $sv->ROOT;
    my ($padlist) = $sv->PADLIST;
    my ($file) = $sv->FILE;
    my ($gv) = $sv->GV;
    printf <<'EOT', $$stash, $$start, $$root;
	STASH		0x%x
	START		0x%x
	ROOT		0x%x
EOT
    if ( $]>5.017 && ($sv->FLAGS & 0x40000)) { #lexsub
      printf("\tNAME\t%%s\n", $sv->NAME);
    } else {
      printf("\tGV\t%0x%x\t%s\n", $$gv, $gv->SAFENAME);
    }
    printf <<'EOT', $file, $sv->DEPTH, $padlist, ${$sv->OUTSIDE};
	FILE		%s
	DEPTH		%d
	PADLIST		0x%x
	OUTSIDE		0x%x
EOT
    printf("\tOUTSIDE_SEQ\t%d\n", $sv->OUTSIDE_SEQ) if $] > 5.007;
    if ($have_B_Flags) {
      my $SVt_PVCV = $] < 5.010 ? 12 : 13;
      printf("\tCvFLAGS\t0x%x\t%s\n", $sv->CvFLAGS,
	     $have_B_Flags_extra ? $sv->flagspv($SVt_PVCV) : $sv->flagspv);
    } else {
      printf("\tCvFLAGS\t0x%x\n", $sv->CvFLAGS);
    }
    $start->debug if $start;
    $root->debug if $root;
    $gv->debug if $gv;
    $padlist->debug if $padlist;
}

sub B::AV::debug {
    my ($av) = @_;
    $av->B::SV::debug;
    _array_debug($av);
}

sub _array_debug {
    my ($av) = @_;
    # tied arrays may leave out FETCHSIZE
    my (@array) = eval { $av->ARRAY; };
    print "\tARRAY\t\t(", join(", ", map("0x" . $$_, @array)), ")\n";
    my $fill = eval { scalar(@array) };
    if ($Config{'useithreads'} && class($av) ne 'PADLIST') {
      printf <<'EOT', $fill, $av->MAX, $av->OFF;
	FILL		%d
	MAX		%d
	OFF		%d
EOT
    } else {
      printf <<'EOT', $fill, $av->MAX;
	FILL		%d
	MAX		%d
EOT
    }
    if ($] < 5.009) {
      if ($have_B_Flags) {
	printf("\tAvFLAGS\t0x%x\t%s\n", $av->AvFLAGS,
	       $have_B_Flags_extra ? $av->flagspv(10) : $av->flagspv);
      } else {
	printf("\tAvFLAGS\t0x%x\n", $av->AvFLAGS);
      }
    }
}

sub B::GV::debug {
    my ($gv) = @_;
    if ($done_gv{$$gv}++) {
	printf "GV %s::%s\n", $gv->STASH->NAME, $gv->SAFENAME;
	return;
    }
    my $sv = $gv->SV;
    my $av = $gv->AV;
    my $cv = $gv->CV;
    $gv->B::SV::debug;
    printf <<'EOT', $gv->SAFENAME, $gv->STASH->NAME, $gv->STASH, $$sv, $gv->GvREFCNT, $gv->FORM, $$av, ${$gv->HV}, ${$gv->EGV}, $$cv, $gv->CVGEN, $gv->LINE, $gv->FILE, $gv->GvFLAGS;
	NAME		%s
	STASH		%s (0x%x)
	SV		0x%x
	GvREFCNT	%d
	FORM		0x%x
	AV		0x%x
	HV		0x%x
	EGV		0x%x
	CV		0x%x
	CVGEN		%d
	LINE		%d
	FILE		%s
EOT
    if ($have_B_Flags) {
      my $SVt_PVGV = $] < 5.010 ? 13 : 9;
      printf("\tGvFLAGS\t0x%x\t%s\n", $gv->GvFLAGS,
	     $have_B_Flags_extra ? $gv->flagspv($SVt_PVGV) : $gv->flagspv);
    } else {
      printf("\tGvFLAGS\t0x%x\n", $gv->GvFLAGS);
    }
    $sv->debug if $sv;
    $av->debug if $av;
    $cv->debug if $cv;
}

sub B::SPECIAL::debug {
    my $sv = shift;
    my $i = ref $sv ? $$sv : 0;
    print defined $specialsv_name[$i] ? $specialsv_name[$i] : "", "\n";
}

sub B::PADLIST::debug {
    my ($padlist) = @_;
    printf <<'EOT', class($padlist), $$padlist, $padlist->REFCNT;
%s (0x%x)
	REFCNT		%d
EOT
    _array_debug($padlist);
}

sub compile {
    my $order = shift;
    B::clearsym();
    if ($order && $order eq "exec") {
        return sub { walkoptree_exec(main_start, "debug") }
    } else {
        return sub { walkoptree(main_root, "debug") }
    }
}

1;

__END__

=head1 NAME

B::Debug - Walk Perl syntax tree, printing debug info about ops

=head1 SYNOPSIS

        perl -MO=Debug foo.pl
        perl -MO=Debug,-exec foo.pl

=head1 DESCRIPTION

See F<ext/B/README> and the newer L<B::Concise>, L<B::Terse>.

=head1 OPTIONS

With option -exec, walks tree in execute order,
otherwise in basic order.

=head1 AUTHOR

Malcolm Beattie, C<mbeattie@sable.ox.ac.uk>
Reini Urban C<rurban@cpan.org>

=head1 LICENSE

Copyright (c) 1996, 1997 Malcolm Beattie
Copyright (c) 2008, 2010, 2013 Reini Urban

	This program is free software; you can redistribute it and/or modify
	it under the terms of either:

	a) the GNU General Public License as published by the Free
	Software Foundation; either version 1, or (at your option) any
	later version, or

	b) the "Artistic License" which comes with this kit.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See either
    the GNU General Public License or the Artistic License for more details.

    You should have received a copy of the Artistic License with this kit,
    in the file named "Artistic".  If not, you can get one from the Perl
    distribution. You should also have received a copy of the GNU General
    Public License, in the file named "Copying". If not, you can get one
    from the Perl distribution or else write to the Free Software Foundation,
    Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA.

=cut

